### PR TITLE
[NSTK-2088] Add initContainer for node-server

### PR DESCRIPTION
**Pull request by sdodsley**
Add initContainer for node-server to pull the node config image which checks iSCSI, multipath and NFS client software is configured and running on the node.
Also applies the Pure Storage udev best practises.

### CODE REVIEW

**Comment by sdodsley** [comment]
Can someone run this through the internal CI please?

**Review by dsupure — APPROVED** [review]
LGTM! I'll ask Adam to kick off a test on the `nod-config` branch before we merge.

**Review by pure-jliao — APPROVED** [review]
so we will have the config setup each time starting a new node right? this is cool!

**Comment by Cellebyte** [comment]
@sdodsley I got the issue that `iscsi-iname` was not random enough i would prefer using `/etc/machine-id`.

**Comment by Cellebyte** [comment]
```bash
#!/bin/sh
nsenter -m/proc/1/ns/mnt -- sh -c '[ -x /usr/bin/dnf ] && echo "Fedora" > /tmp/os-type'
nsenter -m/proc/1/ns/mnt -- sh -c '[ -x /usr/bin/yum ] && echo "RHEL" > /tmp/os-type'
nsenter -m/proc/1/ns/mnt -- sh -c '[ -x /usr/bin/apt ] && echo "Debian" > /tmp/os-type'
OS=$(nsenter -m/proc/1/ns/mnt -- cat /tmp/os-type)
echo ==== Checking node dependencies for PSO ====
if [ $OS = "Debian" ]; then
    nsenter -m/proc/1/ns/mnt -n/proc/1/ns/net -- apt install -y --no-upgrade open-iscsi multipath-tools nfs-common
elif [ $OS = "Fedora" ]; then
    nsenter -m/proc/1/ns/mnt -n/proc/1/ns/net -- dnf install -y --no-upgrade iscsi-initiator-utils device-mapper-multipath nfs-utils libstoragemgmt-udev
    nsenter -m/proc/1/ns/mnt -- sh -c 'mpathconf --enable --with_multipathd y'
elif [ $OS = "RHEL" ]; then
    nsenter -m/proc/1/ns/mnt -n/proc/1/ns/net -- yum install -y iscsi-initiator-utils device-mapper-multipath nfs-utils libstoragemgmt-udev
    nsenter -m/proc/1/ns/mnt -- sh -c 'mpathconf --enable --with_multipathd y'
fi
nsenter -m/proc/1/ns/mnt -- sh -c 'cat <<EOF >>/etc/multipath.conf
defaults {
  polling_interval      10
}
devices {
  device {
      vendor "PURE"
      product "FlashArray"
      fast_io_fail_tmo 10
      path_grouping_policy "group_by_prio"
      failback "immediate"
      prio "alua"
      hardware_handler "1 alua"
      max_sectors_kb 4096
    }
}
EOF'
nsenter -m/proc/1/ns/mnt -- sh -c '[ ! -f /etc/iscsi/initiatorname.iscsi ] && echo "InitiatorName=iqn.2005-03.org.open-iscsi:`cat /etc/machine-id`" > /etc/iscsi/initiatorname.iscsi'
nsenter -m/proc/1/ns/mnt -- sh -c 'if (systemctl status iscsid | grep disabled); then systemctl enable iscsid; fi'
nsenter -m/proc/1/ns/mnt -- sh -c 'if (systemctl status iscsid | grep inactive); then systemctl start iscsid; fi'
nsenter -m/proc/1/ns/mnt -- sh -c 'if (systemctl status multipathd | grep inactive); then systemctl start multipathd; fi'
nsenter -m/proc/1/ns/mnt -- sh -c 'if (systemctl status multipathd | grep inactive); then systemctl start multipathd; fi'
if [ $OS = "Debian" ]; then
    nsenter -m/proc/1/ns/mnt -- sh -c '[ ! -f /lib/udev/rules.d/99-pure-storage.rules ] && cat <<EOF >/etc/udev/rules.d/99-pure-storage.rules
# Recommended settings for Pure Storage FlashArray.
# Use noop scheduler for high-performance solid-state storage for SCSI devices
ACTION=="add|change", KERNEL=="sd*[!0-9]", SUBSYSTEM=="block", ENV{ID_VENDOR}=="PURE", ATTR{queue/scheduler}="none"
ACTION=="add|change", KERNEL=="dm-[0-9]*", SUBSYSTEM=="block", ENV{DM_NAME}=="3624a937*", ATTR{queue/scheduler}="none"
# Reduce CPU overhead due to entropy collection
ACTION=="add|change", KERNEL=="sd*[!0-9]", SUBSYSTEM=="block", ENV{ID_VENDOR}=="PURE", ATTR{queue/add_random}="0"
ACTION=="add|change", KERNEL=="dm-[0-9]*", SUBSYSTEM=="block", ENV{DM_NAME}=="3624a937*", ATTR{queue/add_random}="0"
# Spread CPU load by redirecting completions to originating CPU
ACTION=="add|change", KERNEL=="sd*[!0-9]", SUBSYSTEM=="block", ENV{ID_VENDOR}=="PURE", ATTR{queue/rq_affinity}="2"
ACTION=="add|change", KERNEL=="dm-[0-9]*", SUBSYSTEM=="block", ENV{DM_NAME}=="3624a937*", ATTR{queue/rq_affinity}="2"
# Set the HBA timeout to 60 seconds
ACTION=="add|change", KERNEL=="sd*[!0-9]", SUBSYSTEM=="block", ENV{ID_VENDOR}=="PURE", ATTR{device/timeout}="60"
EOF'
fi
if [ $OS = "RHEL" ] || [ $OS = "Fedora" ]; then
    nsenter -m/proc/1/ns/mnt -- sh -c '[ ! -f /etc/udev/rules.d/99-pure-storage.rules ] && cat <<EOF >/etc/udev/rules.d/99-pure-storage.rules
# Recommended settings for Pure Storage FlashArray.
# Use noop scheduler for high-performance solid-state storage for SCSI devices
ACTION=="add|change", KERNEL=="sd*[!0-9]", SUBSYSTEM=="block", ENV{ID_VENDOR}=="PURE", ATTR{queue/scheduler}="none"
ACTION=="add|change", KERNEL=="dm-[0-9]*", SUBSYSTEM=="block", ENV{DM_NAME}=="3624a937*", ATTR{queue/scheduler}="none"
# Reduce CPU overhead due to entropy collection
ACTION=="add|change", KERNEL=="sd*[!0-9]", SUBSYSTEM=="block", ENV{ID_VENDOR}=="PURE", ATTR{queue/add_random}="0"
ACTION=="add|change", KERNEL=="dm-[0-9]*", SUBSYSTEM=="block", ENV{DM_NAME}=="3624a937*", ATTR{queue/add_random}="0"
# Spread CPU load by redirecting completions to originating CPU
ACTION=="add|change", KERNEL=="sd*[!0-9]", SUBSYSTEM=="block", ENV{ID_VENDOR}=="PURE", ATTR{queue/rq_affinity}="2"
ACTION=="add|change", KERNEL=="dm-[0-9]*", SUBSYSTEM=="block", ENV{DM_NAME}=="3624a937*", ATTR{queue/rq_affinity}="2"
# Set the HBA timeout to 60 seconds
ACTION=="add|change", KERNEL=="sd*[!0-9]", SUBSYSTEM=="block", ENV{ID_VENDOR}=="PURE", ATTR{device/timeout}="60"
EOF'
    nsenter -m/proc/1/ns/mnt -- sh -c 'cat <<EOF >>/etc/udev/rules/90-scsi-ua.rules
ACTION=="change", SUBSYSTEM=="scsi", ENV{SDEV_UA}=="CAPACITY_DATA_HAS_CHANGED", TEST=="rescan", ATTR{rescan}="x"
EOF'
fi
nsenter -m/proc/1/ns/mnt -- sh -c '[ -x /sbin/udevadm ] && /sbin/udevadm control --reload-rules'
nsenter -m/proc/1/ns/mnt -- sh -c '[ -x /sbin/udevadm ] && /sbin/udevadm trigger --type=devices --action=change'
nsenter -m/proc/1/ns/mnt -- sh -c '[ -x /usr/bin/udevadm ] && /usr/bin/udevadm control --reload-rules'
nsenter -m/proc/1/ns/mnt -- sh -c '[ -x /usr/bin/udevadm ] && /usr/bin/udevadm trigger --type=devices --action=change'
exit 0
```

**Comment by sdodsley** [comment]
@Cellebyte can you give examples of where `iscsi-iname` is not random enough? This is the standard way of creating unique IQNs.

**Comment by Cellebyte** [comment]
I think when i boot my Ubuntu from netboot it creates an initiator-iname. Which is the same for every installation.
Then the file exists and the script is not creating a new one.
With the machine-id set you don't need to check for the file being there.
Because the machine-id does not change.

**Comment by sdodsley** [comment]
Do you have a custom install image? From my experience, the standard Ubuntu images do not install iSCSI by default.

**Comment by Cellebyte** [comment]
@sdodsley i install it via ansible and I get the same initiator for all my hosts.

**Comment by sdodsley** [comment]
I would respectfully suggest that you need to reassess your ansible playbook and how it is performing the iSCSI part of the install. A standard install of iSCSI will always create a new initiatorname.iscsi file with a unique IQN.

File: pure-pso/templates/plugin/node-server.yaml
Comment (Pure-AdamuKaapan) [inline pure-pso/templates/plugin/node-server.yaml:46]: Do we want to add something that makes this optional, in case cluster admins don't want packages installed automagically?

**Review by Pure-AdamuKaapan — CHANGES_REQUESTED** [review]
Looking generally good. Blocking for now because we have some cockroach-operator changes that will also have to be factored in here that need to get merged first.

**Comment by Cellebyte** [comment]
> I would respectfully suggest that you need to reassess your ansible playbook and how it is performing the iSCSI part of the install. A standard install of iSCSI will always create a new initiatorname.iscsi file with a unique IQN.

i don't know whats wrong with a apt install open-iscsi.

**Comment by Cellebyte** [comment]
I think using machine-id is much more filesafe.

**Review by haibinxie — APPROVED** [review]
LGTM. Please verify it does not fail if it is already installed.

File: pure-pso/templates/plugin/node-server.yaml
Comment (sdodsley) [reply pure-pso/templates/plugin/node-server.yaml:46]: Interesting idea. Might be a good idea, but the default is to install.

**Comment by sdodsley** [comment]
@Cellebyte  I am confused as to why the `iscsi-name` is creating the same IQN for different hosts. The whole point of this program is to create a unique IQN for each host and is created when open-iscsi is installed. If you have a scenario where this creates the same IQN then this sounds like an issue that should be raised with the open-iscsi team.

**Comment by Cellebyte** [comment]
@sdodsley I know. I am also confused. But I think it has to do with the same start point of entropy for all my installations and its not your script it has to do with the installation of open-iscsi because if your script had created the iqns they would not have debian in them.

File: pure-pso/templates/plugin/node-server.yaml
Comment (sdodsley) [reply pure-pso/templates/plugin/node-server.yaml:47]: Done

**Review by dsupure — APPROVED** [review]
LGTM. We need to make sure it won't conflict with Jenkins's multipath script.

**Review by Pure-AdamuKaapan — CHANGES_REQUESTED** [review]
Overall LGTM, one tiny naming nit. Still blocking because I'll have some cockroach-operator changes for you to put in too, but your side looks good.

File: pure-pso/README.md
Comment (Pure-AdamuKaapan) [reply pure-pso/README.md:185]: Just pushed a commit fixing this, resolving.

**Comment by Pure-AdamuKaapan** [comment]
Updated with the cockroach-operator changes, will unblock once that gets merged too (need to debug latest test run early next week).

**Review by Pure-AdamuKaapan — APPROVED** [review]
LGTM!

File: README.md
Comment (Pure-AdamuKaapan) [inline README.md:39]: Fix this version number (PSO 6.1.1?)

File: pure-pso/templates/plugin/node-server.yaml
Comment (Pure-AdamuKaapan) [inline pure-pso/templates/plugin/node-server.yaml:46]: I believe we have to add the variable to cockroach-operator too, so that it will know to add the init container to the DB pods.

**Review by Pure-AdamuKaapan — CHANGES_REQUESTED** [review]
I think we're missing a change to apply it to cockroach-operator too, but overall pretty close

File: pure-pso/templates/plugin/node-server.yaml
Comment (Pure-AdamuKaapan) [reply pure-pso/templates/plugin/node-server.yaml:46]: Never mind, just realized it's using the same exact variable names so it's fine.

**Review by Pure-AdamuKaapan — APPROVED** [review]
LGTM other than possibly the wrong version number in the README

File: README.md
Comment (dsupure) [reply README.md:39]: @sdodsley : Can you help to update it? Thx

**Comment by dsupure** [comment]
@sdodsley . Thanks.

**Pull request merged**
